Add Lua API docs

IronicallySerious opened the pull request:
Refering to this JSON data: https://gist.github.com/mikke89/05f0a8a8430b53648118340266c69c68
Generated by this Python script: https://gist.github.com/IronicallySerious/9e998798bbf14e7fd4849067527a1cec

mikke89 commented on pages/lua_manual/api_reference.md:
```diff
@@ -0,0 +1,1849 @@
+
+| Return value | Name |
+| ------------ | ---- |
+| `nil`{: .type } | AddColumn(`string`{: .type } fields, `string`{: .type } formatter, `number`{: .type } initial_width, `string`{: .type } header_rml) |
```
```suggestion
| `nil`{: .type } | <a href="#ElementDataGrid-AddColumn" class="lua_method">AddColumn(`string`{: .lua_type } fields, `string`{: .lua_type } formatter, `number`{: .lua_type } initial_width, `string`{: .lua_type } header_rml)</a> |
```
For linking to properties, and so we can style it later. (We could also write the `{:.lua_type}` with html tags if you prefer, I don't mind either way).

IronicallySerious replied:
Great, I am adding this. All table names (class names) are now links and the page links to them from the content table. Also as you did above, I am making all functions and properties to be linkable and their tables also link to them

mikke89 replied:
Great!

mikke89 commented on pages/lua_manual/api_reference.md:
```diff
@@ -0,0 +1,1849 @@
+
+### Property Descriptions
+
+*  `ElementDataGridRow`{: .type } rows
```
```suggestion
* <span id="ElementDataGrid-AddColumn" class="lua_method"><span class="lua_type">ElementDataGridRow</span> rows</span>
```
For linking to it, and styling it.

IronicallySerious replied:
I ended up with getting the same result with this: 
```md
####  `ElementDataGridRow`{: .lua-type } <a name='ElementDataGrid-rows'>rows</a>{: .lua-method }
```

mikke89 replied:
Is this `<a name>` standard html? I'm not sure why we don't just use a span and id?

IronicallySerious replied:
The name attribute in anchor tag also allows us to give the text link formatting. Although I am having second thoughts on this. I ended up giving both name and href attributes to these, where I think name attribute becomes redundant. 
I was initially wanting to give the option to just right click > copy link and share it elsewhere, whether it is a property or a function. Maybe just href would be enough?
I could just do it like you mentioned, with a span and an id but that doesn't allow the user to copy the link directly. They will have to scroll up to the table and then copy the link from there. This can be avoided with we made it a link right there.

mikke89 replied:
I see, copying like that would indeed be useful. I'll leave it to you to figure out :)

IronicallySerious replied:
Well I had to add both `name` and `href` to make it linkable AND to be able to copy the link from it. This should work fine

mikke89 commented on pages/lua_manual/api_reference.md:
```diff
@@ -0,0 +1,1849 @@
+
+| Return value | Name |
+| ------------ | ---- |
+| `nil`{: .type } | AddEventListener(`BOOL`{: .type } event, `string`{: .type } listener[, `lua_type`{: .type } in_capture_phase]) |
```
```suggestion
| `nil`{: .type } | AddEventListener(`boolean`{: .type } event, `string`{: .type } listener[, `lua_type`{: .type } in_capture_phase]) |
```
I think this is the proper Lua terminology?

IronicallySerious replied:
Sure, done

mikke89 commented on pages/lua_manual/api_reference.md:
```diff
@@ -0,0 +1,1849 @@
+| Return value | Name |
+| ------------ | ---- |
+| `nil`{: .type } | Hide() |
+| `nil`{: .type } | Show(`integer`{: .type } [flags]) |
```
This now takes two enums as input.

IronicallySerious replied:
The binding code seems to define extra enums that are supposed to be sent into Document.Show() as arguments.

```cpp
template<> void ExtraInit<Document>(lua_State* L, int metatable_index)
{
    // some stuff

    //create the DocumentFocus table
    lua_getglobal(L,"DocumentModal");
    if(lua_isnoneornil(L,-1))
    {
        lua_pop(L,1); //pop unsucessful getglobal
        lua_newtable(L); //create a table for holding the enum
        lua_pushinteger(L,(int)ModalFlag::None);
        lua_setfield(L,-2,"None");
        lua_pushinteger(L,(int)ModalFlag::Modal);
        lua_setfield(L,-2,"Modal");
		lua_pushinteger(L, (int)ModalFlag::Keep);
        lua_setfield(L,-2,"Keep");
        lua_setglobal(L,"DocumentModal");
    }

	//create the DocumentFocus table
	lua_getglobal(L, "DocumentFocus");
	if (lua_isnoneornil(L, -1))
	{
		lua_pop(L, 1); //pop unsucessful getglobal
		lua_newtable(L); //create a table for holding the enum
		lua_pushinteger(L, (int)FocusFlag::None);
		lua_setfield(L, -2, "None");
		lua_pushinteger(L, (int)FocusFlag::Document);
		lua_setfield(L, -2, "Document");
		lua_pushinteger(L, (int)FocusFlag::Keep);
		lua_setfield(L, -2, "Keep");
		lua_pushinteger(L, (int)FocusFlag::Auto);
		lua_setfield(L, -2, "Auto");
		lua_setglobal(L, "DocumentFocus");
	}
}
```

And depending on how many arguments are passed into Show(), the correct Show() overload is called by the bindings:

```cpp
int DocumentShow(lua_State* L, Document* obj)
{
    int top = lua_gettop(L);
    if(top == 0)
        obj->Show();
    else if(top == 1)
    {
        ModalFlag modal = (ModalFlag)luaL_checkinteger(L,1);
        obj->Show(modal);
    }
	else
	{
        ModalFlag modal = (ModalFlag)luaL_checkinteger(L,1);
		FocusFlag focus = (FocusFlag)luaL_checkinteger(L,2);
		obj->Show(modal, focus);
	}
    return 0;
}
```

These sort of things might have been missed in the JSON

IronicallySerious replied:
I think there may be more problems like this in the API. I will try to fix all these once on my own by going through the bindings manually.

mikke89 replied:
Yeah, I think this needs to be fixed in manual review. 

Also, everything which has `lua_type` arguments means that you can pass in values of different types which will be resolved in the function (different function signatures if you will). You don't actually pass a type. These ones will probably need manual review as well.

mikke89 reviewed (commented):
Thanks a lot!

Let me know what you think of my suggestions. A bit of manual revision as well, and it should be good to go :)

IronicallySerious commented on pages/lua_manual/api_reference.md:
```diff
@@ -0,0 +1,1849 @@
+All instantiable classes define a `new()` method which returns an object of that particular class. 
+
+With the exception of this `new()` method, all members listed will be member methods.
+
```
Yeah it would be very helpful. Adding this

IronicallySerious commented on pages/lua_manual/api_reference.md:
```diff
@@ -0,0 +1,1849 @@
+
+### Properties
+
+| Types | Name |
```
Would we want the properties to have a type or do we want them to return a value? I think properties should be differentiated from functions based on this.

And yeah putting the name before the type is fine

IronicallySerious replied:
I am changing `Types` to `Type` for now, till this is decided

mikke89 replied:
I'm fine with this :)

IronicallySerious commented on pages/lua_manual/api_reference.md:
```diff
@@ -0,0 +1,1849 @@
+
+### Methods
+
+| Return value | Name |
```
I think you mean something like this:

|Member|Type|
|------|----|
|some_property|Property|
|int SomeMethod(int param1, int param2)|Method|
|__add|Metafunction|

This could get a lot crowded when there are many function arguments and long return types. And to fix this we could just mention the function name and nothing else, but that is kind of going against the motive here I think.

mikke89 replied:
Alright, I'll trust you with this, looks pretty good now.

mikke89 commented on pages/lua_manual/api_reference.md:
```diff
@@ -0,0 +1,1961 @@
+
+### Methods
+
+| Return Type | Name |
```
Maybe put the return types at the end here as well, for consistency with the properties?

IronicallySerious replied:
Alright will do this

IronicallySerious replied:
Actually I also now want to change "Method" to "Function" everywhere. The term sort of carried over from the Python terminology and Lua uses the word 'function' for it

mikke89 replied:
Sounds good to me.

IronicallySerious commented:
Alright, the formatting should be okay now. If there are any more changes to that let me know. I will now go through the binding code and try to verify all the docstrings.

mikke89 commented:
Looks good to me. The only thing that stands out a bit to me is the long table of contents list. Maybe we could section it a bit better. For example "main types" such as LuaRmlUi, Context, Element. And `Utilities`: Vector, Color. `Special elements`: ElementFormControlInput etc. These are just suggestions.

IronicallySerious commented:
Sure that will be nice. I can break the table into those 3 types manually.

IronicallySerious commented:
@mikke89 Sent the first manual review I was able to do. I went over the bindings in the Lua plugin and verified all of them with the docs

IronicallySerious commented:
I will be doing another review over all the docs strings to verify that all of them make sense

mikke89 commented:
Nice work, let me know when you feel it's ready to be merged.

By the way, are you still generating the markdown from your script? If so, it would be nice to include the script and json data into the repository so that it is easy to extend and modify the documentation in the future.

IronicallySerious commented:
@mikke89 I have gone over the docstrings once and fixed all the "Python terminology" that I could find (which wasn't translating to Lua as it is very well) and converted them to their Lua counterparts.

It should be okay to merge from my side.

And no, after [this commit](https://github.com/mikke89/RmlUiDoc/pull/4/commits/76664eb69c9e33a7880516c183c843ca22c464c6), I started doing manual changes just because I didn't want to edit the JSON directly since that was a single source of truth I had and the task left was the manual review itself. But anyways, I have now pushed the script and the data in this PR so that it can be used again later if required.

Although I don't see it being used later unless the docs become *too* out of date with the original API :)

mikke89 commented:
Alright, thanks a lot!

I figured the script might be useful when we add or modify functions, or perhaps want to change the layout later. At least now we have it just in case it becomes useful :)

I'll merge this very soon. Do you mind if I put the RmlUi MIT license on the python script?

IronicallySerious commented:
Sure go ahead :)

On Tue, Sep 22, 2020, 16:32 Michael R. P. Ragazzon <notifications@github.com>
wrote:

> Alright, thanks a lot!
>
> I figured the script might be useful when we add or modify functions, or
> perhaps want to change the layout later. At least now we have it just in
> case it becomes useful :)
>
> I'll merge this very soon. Do you mind if I put the RmlUi MIT license on
> the python script?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mikke89/RmlUiDoc/pull/4#issuecomment-696650598>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHSUAFBTF7R7WEK2UGR2H23SHB73BANCNFSM4RTIXX5Q>
> .
>

mikke89 commented:
Thanks again! This has been a long time on the to-do list.
